lib/auditd/auditd_linux.go

```diff
@@ -0,0 +1,353 @@
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (c *Client) connectUnderMutex() error {
+	if c.conn != nil {
+		// Already connected, return
+		return nil
+	}
+
+	conn, err := c.dial(syscall.NETLINK_AUDIT, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	c.conn = conn
+
+	return nil
+}
+
+func (c *Client) isEnabledUnderMutex() (bool, error) {
+	if c.enabled != unset {
+		// We've already gotten the status.
+		return c.enabled == enabled, nil
+	}
+
+	status, err := getAuditStatus(c.conn)
+	if err != nil {
+		return false, trace.Errorf("failed to get auditd status: %v", trace.ConvertSystemError(err))
+	}
+
+	// enabled can be either 1 or 2 if enabled, 0 otherwise
+	if status.Enabled > 0 {
+		c.enabled = enabled
+	} else {
+		c.enabled = disabled
+	}
+
+	return c.enabled == enabled, nil
+}
+
+// NewClient creates a new auditd client. Client is not connected when it is returned.
+func NewClient(msg Message) *Client {
+	msg.SetDefaults()
+
+	execName, err := os.Executable()
+	if err != nil {
+		log.WithError(err).Warn("failed to get executable name")
+		execName = UnknownValue
+	}
+
+	// Teleport never tries to get the hostname name.
+	// Let's mimic the sshd behavior.
+	const hostname = UnknownValue
+
+	return &Client{
+		execName:     execName,
+		hostname:     hostname,
+		systemUser:   msg.SystemUser,
+		teleportUser: msg.TeleportUser,
+		address:      msg.ConnAddress,
+		ttyName:      msg.TTYName,
+
+		dial: func(family int, config *netlink.Config) (NetlinkConnector, error) {
+			return netlink.Dial(family, config)
+		},
+	}
+}
+
+func getAuditStatus(conn NetlinkConnector) (*auditStatus, error) {
+	_, err := conn.Execute(netlink.Message{
+		Header: netlink.Header{
+			Type:  netlink.HeaderType(AuditGet),
+			Flags: netlink.Request | netlink.Acknowledge,
+		},
+		Data: nil,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	msgs, err := conn.Receive()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(msgs) != 1 {
+		return nil, trace.Errorf("returned wrong messages number, expected 1, got: %d", len(msgs))
+	}
+
+	// auditd marshalling depends on the system architecture.
+	byteOrder := nlenc.NativeEndian()
+	status := &auditStatus{}
+
+	payload := bytes.NewReader(msgs[0].Data[:])
+	if err := binary.Read(payload, byteOrder, status); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return status, nil
+}
+
+// SendMsg sends a message. Client will create a new connection if not connected already.
+func (c *Client) SendMsg(event EventType, result ResultType) error {
+	const msgDataTmpl = `op={{ .Opcode }} acct="{{ .Msg.SystemUser }}" exe={{ .Exe }} ` +
+		`hostname={{ .Hostname }} addr={{ .Msg.ConnAddress }} terminal={{ .Msg.TTYName }} ` +
+		`{{if .Msg.TeleportUser}}teleportUser={{.Msg.TeleportUser}} {{end}}res={{ .Result }}`
+	op := eventToOp(event)
+
+	buf := &bytes.Buffer{}
+	if err := template.Must(template.New("auditd-message").
+		Parse(msgDataTmpl)).
```

Comment:
Can `template.Must(template.New("...").Parse(...))` be moved to a global scope variable so this function just executes it?